napytau/core/delta_tau.py

```diff
@@ -0,0 +1,148 @@
+    jacobian matrix and a weight matrix derived from the shifted intensities' errors.
+    Args:
+        delta_shifted_intensities (ndarray): Errors in the shifted intensities.
+        distances (ndarray): Array of distance points.
+        coefficients (ndarray): Array of polynomial coefficients.
+
+    Returns:
+        ndarray: The computed covariance matrix for the polynomial coefficients.
+    """
+
+    # Compute the Jacobian matrix for the polynomial
+    jacobian_matrix: ndarray = calculate_jacobian_matrix(distances, coefficients)
+
+    # Construct the weight matrix from the inverse squared errors
+    weight_matrix: ndarray = diag(1 / power(delta_shifted_intensities, 2))
+
+    # Compute the fit matrix
+    fit_matrix: ndarray = jacobian_matrix.T @ weight_matrix @ jacobian_matrix
+
+    # Invert the fit matrix to get the covariance matrix
+    covariance_matrix: ndarray = linalg.inv(fit_matrix)
+
+    return covariance_matrix
+
+
+def calculate_error_propagation_terms(
+    unshifted_intensities: ndarray,
+    delta_shifted_intensities: ndarray,
+    delta_unshifted_intensities: ndarray,
+    distances: ndarray,
+    coefficients: ndarray,
+    taufactor: float,
+) -> ndarray:
+    """
+    creates the error propagation term for the polynomial coefficients.
+    combining direct errors, polynomial uncertainties, and mixed covariance terms.
+    Args:
+        unshifted_intensities (ndarray): Unshifted intensity values.
+        delta_shifted_intensities (ndarray): Errors in the shifted intensities.
+        delta_unshifted_intensities (ndarray): Errors in the unshifted intensities.
+        distances (ndarray): Array of distance points.
+        coefficients (ndarray): Array of polynomial coefficients.
+        taufactor (float): Scaling factor related to the Doppler-shift model.
+
+    Returns:
+        ndarray: The combined error propagation terms for each time point.
+    """
+
+    calculated_differentiated_polynomial_sum_at_measuring_times = (
+        differentiated_polynomial_sum_at_measuring_distances(  # noqa E501
+            distances,
+            coefficients,
+        )
+    )
+
+    # First summand: Contribution from unshifted intensity errors
+    first_summand: ndarray = power(delta_unshifted_intensities, 2) / power(
+        calculated_differentiated_polynomial_sum_at_measuring_times,
+        2,
+    )
+
+    # Initialize the polynomial uncertainty term for second term
+    delta_p_j_i_squared: ndarray = zeros(len(distances))
+    covariance_matrix: ndarray = calculate_covariance_matrix(
+        delta_shifted_intensities, distances, coefficients
+    )
+
+    # Calculate the polynomial uncertainty contributions
+    for k in range(len(coefficients)):
+        for l in range(len(coefficients)):  # noqa E741
+            delta_p_j_i_squared = (
+                delta_p_j_i_squared
+                + power(distances, k) * power(distances, l) * covariance_matrix[k, l]
```

Comment:
We should probably check that the dimensions always match. It should be true, because the shape of the jacobian_matrix is (len(distances), len(coefficients)), the shape of the weight_matrix is (len(coefficients), len(distances) and so the shape of the fit_matrix should always be (len(coefficients), len(coefficients)).